¡ Hola Jorge! Como te va?

Mi nombre es Facundo Lozano! Ya he tenido el agrado de revisar otros proyectos tuyos, nuevamente seré tu revisor en este proyecto.

Como siempre, a continuación un poco sobre la modalidad de revisión que usaremos:

Cuando enccuentro un error por primera vez, simplemente lo señalaré, te dejaré encontrarlo y arreglarlo tú cuenta. Además, a lo largo del texto iré haciendo algunas observaciones sobre mejora en tu código y también haré comentarios sobre tus percepciones sobre el tema. Pero si aún no puedes realizar esta tarea, te daré una pista más precisa en la próxima iteración y también algunos ejemplos prácticos. Estaré abierto a comentarios y discusiones sobre el tema.

Encontrará mis comentarios a continuación: no los mueva, modifique ni elimine.

Puedes encontrar mis comentarios en cuadros verdes, amarillos o rojos como este:


<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Exito. Todo se ha hecho de forma exitosa.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Observación. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Necesita arreglos. Este apartado necesita algunas correcciones. El trabajo no puede ser aceptado con comentarios rojos. 
</div>

Puedes responder utilizando esto:

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>
</div>

# Rusty Bargain 

El servicio de venta de autos usados Rusty Bargain está desarrollando una aplicación para atraer nuevos clientes. Gracias a esa app, puedes averiguar rápidamente el valor de mercado de tu coche. Tienes acceso al historial: especificaciones técnicas, versiones de equipamiento y precios. Tienes que crear un modelo que determine el valor de mercado.

# Tabla de contenido


1. [Introducción](#Introducción)

    1.1 [Objetivo general](#Objetivo-general)
    
    1.2 [Objetivos particulares](#Objetivos-particulares)
    
    
2. [Descripción de los datos](#Descripción-de-los-datos)

    2.1 [Datos utilizados como características](#Datos-utilizados-como-características)
    
    2.2 [Datos utilizados como objetivo](#Datos-utilizados-como-objetivo)


3. [Preprocesamiento de datos](#Preprocesamiento-de-datos)

    3.1 [Limpieza de datos](#Limpieza-de-datos)
    
    3.2 [Preparación de datos](#Preparación-de-datos)


4. [Entrenamiento del modelo](#Entrenamiento-del-modelo)

    4.1 [Entrenamiento con modelo de regresión lineal](#Entrenamiento-con-modelo-de-regresión-lineal)
    
    4.2 [Entrenamiento con regresión lineal con descenso de gradiente estocástico](#Entrenamiento-con-regresión-lineal-con-descenso-de-gradiente-estocástico)
    
    4.3 [Entrenamiento de modelo mediante LightGBM](#Entrenamiento-de-modelo-mediante-LightGBM)


5. [Análisis del modelo](#Análisis-del-modelo)

    5.1 [Analizando pruebas con el modelo de regresión lineal SGDL](#Analizando-pruebas-con-el-modelo-de-regresión-lineal-SGDL)
    
    5.2 [Analizando pruebas con el modelo de regresión lineal GBDT](#Analizando-pruebas-con-el-modelo-de-regresión-lineal-GBDT)  
    
    
6. [Conclusión](#Conclusión)


7. [Lista de control](#Lista-de-control)

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 1) </b> <a class="tocSkip"></a>

Jorge, como siempre, me tomo este tiempo al inicio del proyecto para comentar mis apreciaciones generales de esta primera iteración de la entrega. 

Siempre me gusta comenzar dando la bienvenida al mundo de los datos a los estudiantes, te deseo lo mejor y espero que consigas lograr tus objetivos. Personalmente siempre me gusta brindar el siguiente consejo, "Está bien equivocarse, es normal y es lo mejor que te puede pasar. Aprendemos de los errores y eso te hará mejor programando ya que podrás descubrir cosas a medida que avances y son estas cosas las que te darán esa experiencia para ser mejor como Data Scientist"
    
Ahora si yendo a esta notebook.  Jorge, lo he dicho al final del proyecto pero lo resalto aquí nuevamente, tu proyecto está muy bien resuelto, resalta capacidad y comprensión de todas las herramientas, como a la vez esta ordenado y es sencillo de seguir junto a agregados fantasticos, felictiaciones!

Este proyecto está en condiciones de ser aprobado! Éxitos dentro de tu camino en el mundo de los datos!

Saludos Jorge!

## Introducción

### Objetivo general (Rusty Bargain)

- Desarrollar un modelo de predicción para evaluar el valor en el mercado sobre las característcias del coche de cualquier cliente usuario para la nueva app de la empresa.

### Objetivos particulares(Rusty Bargain)

- Que el modelo cumpla con una alta calidad de la predicción.
- Obtener una buena velocidad en la predicción.
- Que el tiempo requerido para el entrenamiento sea el menor posible.



### Instalar lo que falte.

In [ ]:
# instalar librerías que se van a utiliza
!pip install lightgbm

### importar librerias que se van a utilizar

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## para la imputación de valores ausentes
#from sklearn.impute import SimpleImputer

## para la estandarización de datos
from sklearn.preprocessing import StandardScaler

## para la división y subconjunto de datos (train y test)
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
## para importar métricas de evaluación
from sklearn.metrics import make_scorer, r2_score, mean_squared_error

## para medir el tiempo que tarda en ejecutarse un proceso
import time

## para hacer modelos de aprendizaje GBDT
import lightgbm as lgb

## para crear modelo de regresión lineal
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import SGDRegressor
## para graficar
import matplotlib.pyplot as plt

### importar la base de datos que se va a utilizar


In [ ]:
try:
    df = pd.read_csv('/datasets/car_data.csv')
except:
    df = pd.read_csv('car_data.csv')

In [ ]:
print(df.info())

In [ ]:
df.head()

In [ ]:
df.describe()


<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
Excelente Jorge, hasta el momento hemos cargado correctamente los datos separandolos de las importaciones para mitigar posibles errores, a la vez has implementado métodos para visualizar y comprender la composición de nuestros datos. Bien hecho!

### Descripcion de las Columnas 

- **DateCrawled** — *fecha en la que se descargó el perfil de la base de datos*
-  **VehicleType** — *tipo de carrocería del vehículo*
- **RegistrationYear** — *año de matriculación del vehículo*
- **Gearbox** — *tipo de caja de cambios*
- **Power** — *potencia (CV)*
- **Model** — *modelo del vehículo*
- **Mileage** — *kilometraje (medido en km de acuerdo con las especificidades regionales del conjunto de datos)*
- **RegistrationMonth** — *mes de matriculación del vehículo*
- **FuelType** — *tipo de combustible*
- **Brand** — *marca del vehículo*
- **NotRepaired** — *vehículo con o sin reparación*
- **DateCreated** — *fecha de creación del perfil*
- **NumberOfPictures** — *número de fotos del vehículo*
- **PostalCode** — *código postal del propietario del perfil (usuario)*
- **LastSeen** — *fecha de la última vez que el usuario estuvo activo*

### Tipo de moneda

- **Price** — *precio (en euros)*

## Preparación de datos

### Limpieza de datos
Para optimizar el análisis y garantizar mayor claridad en los resultados, hemos decidido reducir el volumen de datos eliminando algunas columnas que, por razones lógicas, no serán relevantes para nuestros objetivos.

In [ ]:
# eliminando columnas de poco valor para el entrenamiento del modelo
low_value_features = ['DateCrawled','RegistrationMonth','DateCreated','NumberOfPictures','PostalCode','LastSeen']
df.drop(low_value_features, axis=1, inplace=True)

In [ ]:
# aplicando el formato "sneake_case" a los nombres de las columnas
rename_columns = {
    'Price' : 'price',
    'VehicleType' : 'vehicle_type',
    'RegistrationYear' : 'registration_year',
    'Gearbox' : 'gearbox',
    'Power' : 'power',
    'Model' : 'model',
    'Mileage' : 'mileage',
    'FuelType' : 'fuel_type',
    'Brand' : 'brand',
    'NotRepaired' : 'not_repaired'
}
df.rename(columns=rename_columns, inplace=True)

In [ ]:
df.info()

Se observa que la tipografía en cada uno de los datos es la correcta por lo que no es necesario hacer correcciones de carácter tipográfico.

In [ ]:
# calculo de porcentaje de pérdida en caso de eliminar los valores ausentes
df_sinvaloresnulos = (df.dropna().shape[0])/(df.shape[0])
print(f'La cantidad de observaciones que quedan después de eliminar los valores ausentes es de {df_sinvaloresnulos :.1%}.')

<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
Un gran avance! Excelente decisión por un lado la de eliminar features que no aportan valor al proyecto, por normalizar los nombres de las features para ser más ordenados y por ver cuanto representa el quitar aquellos valores ausentes en el proyecto, impreisonante!

Dado el alto nivel de pérdida de datos que se podría experimentar, adoptaremos la estrategia de imputar los valores ausentes. Aunque esto podría introducir un pequeño sesgo respecto a los valores reales, preservar la mayor cantidad de información posible permitirá obtener mejores predicciones al entrenar nuestros modelos.

In [ ]:
# observar las estadísticas primarias de las columnas con valores cuantitativos
df.describe()

In [ ]:
# Configura el estilo de Seaborn para mejorar la estética
sns.set(style="whitegrid")

plt.figure(figsize=(9,9))

# Gráfica 1: Diagrama de caja para 'power'
plt.subplot(2, 1, 1)
sns.boxplot(data=df, y='power', color='lightblue')
plt.title('Distribución de Power', fontsize=14)

# Gráfica 2: Diagrama de caja para 'registration_year'
plt.subplot(2, 1, 2)
sns.boxplot(data=df, y='registration_year', color='lightgreen')
plt.title('Distribución de Registration Year', fontsize=14)

plt.tight_layout()  # Ajusta el espacio entre las gráficas
plt.show()

Existen muchos valores atípicos por lo cual se procederá a eliminarlos

In [ ]:
# Creamos una función que elimine valores atípicos dependientes de columna

def drop_av(df, column):
    temp = df[column]
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3-Q1
    bottom_limit = Q1-(1.5*IQR)
    upper_limit = Q3+(1.5*IQR)
    df = df.query('@temp > @bottom_limit and @temp < @upper_limit')
    return df

In [ ]:
# se ejecuta la función anterior
df = drop_av(df, 'power')
df = drop_av(df, 'registration_year')

In [ ]:
# Configura el estilo de Seaborn para mejorar la estética
sns.set(style="whitegrid")

plt.figure(figsize=(10,10))

# Gráfica 1: Diagrama de caja para 'power'
plt.subplot(2, 1, 1)
sns.boxplot(data=df, y='power', color='green')
plt.title('Distribución de Power', fontsize=14)
plt.ylabel('Power', fontsize=12)

# Gráfica 2: Diagrama de caja para 'registration_year'
plt.subplot(2, 1, 2)
sns.boxplot(data=df, y='registration_year', color='red')
plt.title('Distribución de Registration Year', fontsize=14)
plt.ylabel('Registration Year', fontsize=12)

# Ajusta el espaciado entre las gráficas
plt.tight_layout()
plt.show()

Se ha identificado que los valores ausentes están presentes únicamente en las columnas con características cualitativas, mientras que las características cuantitativas no presentan valores faltantes. Para manejar estos datos categóricos, podríamos aplicar la codificación One-Hot Encoding (OHE). Aunque columnas como `model` o `brand` generarán un gran número de nuevas columnas, es fundamental que la base de datos reconozca estas características, ya que son determinantes en la predicción del precio y su impacto en el modelo será significativo.

In [ ]:
## corrigiendo el comentario anterior
data = df.fillna('unknown')

<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
Nuevamente excelente Jorge! Siempre es importante  analizar los outliers en nuestros problemas, en este caso teniamos casos con años de registro claramente erroneos como 9999, bien implementado y bien mostrado mediante el uso de boxplots! Excelente uso de los quantiles para este acercamiento :)

In [ ]:
## convirtiendo las variables categóricas en tu DataFrame df en variables binarias (o "dummies"). Esto se conoce como One-Hot Encoding.
data = pd.get_dummies(df, drop_first=True)

### Preparación de datos

In [18]:
# se van a manejar lo datos de manera estandarziada
to_std = ['registration_year','power','mileage']
scaler = StandardScaler()
scaler.fit(data[to_std])
data_scaled = data
data_scaled[to_std]= scaler.transform(data_scaled[to_std])
data_scaled.head()

price  registration_year     power   mileage  vehicle_type_convertible  \
0    480          -1.654579 -1.826334  0.569253                         0   
1  18300           1.160785  1.539047 -0.100732                         0   
2   9800           0.065921  1.060809 -0.100732                         0   
3   1500          -0.403306 -0.497894  0.569253                         0   
4   3600           0.691557 -0.604169 -1.038711                         0   

   vehicle_type_coupe  vehicle_type_other  vehicle_type_sedan  \
0                   0                   0                   0   
1                   1                   0                   0   
2                   0                   0                   0   
3                   0                   0                   0   
4                   0                   0                   0   

   vehicle_type_small  vehicle_type_suv  ...  brand_skoda  brand_smart  \
0                   0                 0  ...            0            0   
1                   0                 0  ...            0            0   
2                   0                 1  ...            0            0   
3                   1                 0  ...            0            0   
4                   1                 0  ...            1            0   

   brand_sonstige_autos  brand_subaru  brand_suzuki  brand_toyota  \
0                     0             0             0             0   
1                     0             0             0             0   
2                     0             0             0             0   
3                     0             0             0             0   
4                     0             0             0             0   

   brand_trabant  brand_volkswagen  brand_volvo  not_repaired_yes  
0              0                 1            0                 0  
1              0                 0            0                 1  
2              0                 0            0                 0  
3              0                 1            0                 0  
4              0                 0            0                 0  

[5 rows x 304 columns]

<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
Excelente implementación de get_dummies y StandarScaler para la transformación de nuestras features tanto de categóricas a numéricas como numéricas normalizadas. Sigamos!

In [19]:
# crear el conjunto de entrenamiento y de prueba

X_train, X_test, y_train, y_test = train_test_split(data_scaled.drop(['price'], axis=1), data_scaled['price'], random_state=300623, test_size=0.25)

## Entrenamiento del modelo 
### Entrenamiento con modelo de regresión lineal

In [20]:
t0 = time.time()
rl_model = LinearRegression()
rl_model.fit(X_train, y_train)
y_pred = rl_model.predict(X_test)
t1 = time.time()
total_time = t1 - t0
mse = mean_squared_error(y_pred, y_test)
rmse = np.sqrt(mse)
print(f' r2_score: {r2_score(y_pred, y_test)}.')
print(f' RECM: {rmse}.')
print(f' Tiempo total de entrenamiento: {total_time}.')

 r2_score: 0.4756784519145183.
 RECM: 2565.024252882704.
 Tiempo total de entrenamiento: 11.826346397399902.


<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
Modelo simple por default para ver un acercamiento a posibles resultados, bien implementado y excelente uso de métricas y medición de tiempo!

Se puede observar que el modelo es extremandamente deficiente.

### Entrenamiento con regresión lineal con descenso de gradiente estocástico

In [21]:
class SGDLinearRegression:
    def __init__(self, step_size, epochs, batch_size, reg_weight):
        self.step_size = step_size
        self.epochs = epochs
        self.batch_size = batch_size
        self.reg_weight = reg_weight
    
    def fit(self, train_features, train_target):
        X = np.concatenate(
            (np.ones((train_features.shape[0], 1)), train_features), axis=1
        )
        y = train_target
        w = np.zeros(X.shape[1]).astype('float64')
        
        for _ in range(self.epochs):
            batches_count = X.shape[0] // self.batch_size
            for i in range(batches_count):
                begin = i * self.batch_size
                end = (i + 1) * self.batch_size
                X_batch = X[begin:end, :]
                y_batch = y[begin:end]
                
                gradient = (2 * X_batch.T.dot(X_batch.dot(w) - y_batch)/ X_batch.shape[0]).astype('float64')
                
                reg = 2 * w.copy()
                reg[0] = 0
                
                gradient += self.reg_weight * reg
                
                w -= self.step_size*gradient

        self.w = w[1:]
        self.w0 = w[0]

    def predict(self, test_features):
        return test_features.dot(self.w) + self.w0

In [22]:
## entrenamineto del modelo SGDRL
    
t0 = time.time()    
model = SGDLinearRegression(0.01, 10, 100, 0)
model.fit(X_train, y_train)
pred_test = model.predict(X_test)
t1 = time.time()
mse = mean_squared_error(pred_test, y_test)
rmse = np.sqrt(mse)
process_time = (t1-t0)
print(f' métrica r2_score: {r2_score(y_test, pred_test).round(4)}.')
print(f' RECM: {rmse}.')
print(f'Tiempo de entrenamiento y predicción es:{process_time} segundos.')


 métrica r2_score: 0.6433.
 RECM: 2610.15363570392.
Tiempo de entrenamiento y predicción es:70.20427775382996 segundos.


<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
Impresionante armado manual  para la regresión lineal teniendo en cuenta el descenso de gradiente Jorge! Excelente trabajo!

### Entrenamiento de modelo mediante LightGBM

In [23]:
train_data = lgb.Dataset(X_train, label=y_train)
test_data = lgb.Dataset(X_test, label=y_test)

In [24]:
## especificando parámetros a utilizar para lightgbm

parameters = {
    'force_row_wise' : True,
    'is_unbalance' : True,
    'boosting_type' : 'gbdt',
    'verbosity' : -1,
    'num_leaves' : 150,
    'learning_rate' : 0.3,
    'metric' : ['rmse', 'auc'],
    'random_state' : 1266,
    'subsample_for_bin' : 10000,
    'n_estimators' : 300
}

In [25]:
callbacks = [lgb.early_stopping(10)]
t0 = time.time()
model_gbm = lgb.train(parameters,                      
                      train_data,
                      valid_sets = test_data,
                      num_boost_round =1000,
                      callbacks = callbacks,
                     )
predict = model_gbm.predict(X_test)
t1 = time.time()
process_time = (t1-t0)
print(f'Tiempo de entrenamiento y predicción es:{process_time} segundos.')

/opt/conda/envs/python3/lib/python3.9/site-packages/lightgbm/engine.py:177: UserWarning: Found `n_estimators` in params. Will use it instead of argument
  _log_warning(f"Found `{alias}` in params. Will use it instead of argument")


Training until validation scores don't improve for 10 rounds
Early stopping, best iteration is:
[72]	valid_0's rmse: 1600.38	valid_0's auc: 0.702606
Tiempo de entrenamiento y predicción es:16.381969213485718 segundos.


El proceso anterior muestra un excelente `RECM` y un `AUC` deseable; sin embargo, el tiempo de entrenamiento, que es de aproximadamente 8.6 minutos con los recursos de procesamiento actuales, es relativamente alto. Por lo tanto, es recomendable optimizar los hiperparámetros del modelo para mejorar el rendimiento y reducir el tiempo de entrenamiento.

In [26]:
mse = mean_squared_error(predict, y_test)
rmse = np.sqrt(mse)
print(f'Métrica r2_score: {r2_score(y_test, predict).round(4)}.')
print(f'RECM: {rmse}.')

Métrica r2_score: 0.8659.
RECM: 1600.3756623334853.


<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
 
Y un tercer modelo en este caso utilizando una grilla para los parametros, bien hecho y buenos resultados!

## Análisis del modelo
### Analizando pruebas con el modelo de regresión lineal SGDL

In [27]:
t0 = time.time()
model =SGDLinearRegression(0.01, 10, 100, 0.1)
model.fit(X_train, y_train)
pred_test = model.predict(X_test)
t1 = time.time()
process_time = (t1-t0)
print(r2_score(y_test, pred_test).round(4))
print(f'Tiempo de entrenamiento y predicción es:{process_time} segundos.')

0.5608
Tiempo de entrenamiento y predicción es:71.70253658294678 segundos.


In [28]:
t0 = time.time()
model =SGDLinearRegression(0.01, 10, 100, 1)
model.fit(X_train, y_train)
pred_test = model.predict(X_test)
t1 = time.time()
process_time = (t1-t0)
print(r2_score(y_test, pred_test).round(4))
print(f'Tiempo de entrenamiento y predicción es:{process_time} segundos.')

0.4166
Tiempo de entrenamiento y predicción es:73.09981894493103 segundos.


In [29]:
t0 = time.time()
model =SGDLinearRegression(0.01, 10, 100, 10)
model.fit(X_train, y_train)
pred_test = model.predict(X_test)
t1 = time.time()
process_time = (t1-t0)
print(r2_score(y_test, pred_test).round(4))
print(f'Tiempo de entrenamiento y predicción es:{process_time} segundos.')

0.1028
Tiempo de entrenamiento y predicción es:71.87703800201416 segundos.


In [30]:
t0 = time.time()
model =SGDLinearRegression(0.1, 10, 1000, 0)
model.fit(X_train, y_train)
pred_test = model.predict(X_test)
t1 = time.time()
process_time = (t1-t0)
print(r2_score(y_test, pred_test).round(4))
print(f'Tiempo de entrenamiento y predicción es:{process_time} segundos.')

0.6432
Tiempo de entrenamiento y predicción es:14.882035493850708 segundos.


### Analizando pruebas con el modelo de regresión lineal GBDT

Así nos referimos al modelo realizado con las herramientas de LightGBM en el cual se utilizó potenciación del gradiente con árboles de decisión.
Se puede observar que se obtiene el mejor rendimiento en el modelo.

In [31]:
train_data = lgb.Dataset(X_train, label=y_train)
test_data = lgb.Dataset(X_test, label=y_test)

In [32]:
callbacks = [lgb.early_stopping(10)]
t0 = time.time()
model_gbm_2 = lgb.LGBMClassifier()
model_gbm_2 = lgb.train(parameters,
                      train_data,
                      valid_sets = test_data,
                      num_boost_round =500,
                      callbacks = callbacks
                     )
predict = model_gbm_2.predict(X_test)
t1 = time.time()
process_time = (t1-t0)
print(f'Tiempo de entrenamiento y predicción es:{process_time} segundos.')

/opt/conda/envs/python3/lib/python3.9/site-packages/lightgbm/engine.py:177: UserWarning: Found `n_estimators` in params. Will use it instead of argument
  _log_warning(f"Found `{alias}` in params. Will use it instead of argument")


Training until validation scores don't improve for 10 rounds
Early stopping, best iteration is:
[72]	valid_0's rmse: 1600.38	valid_0's auc: 0.702606
Tiempo de entrenamiento y predicción es:17.6119601726532 segundos.


In [33]:
mse = mean_squared_error(predict, y_test)
rmse = np.sqrt(mse)
print(f'Métrica r2_score: {r2_score(y_test, predict).round(4)}.')
print(f'RECM: {rmse}.')

Métrica r2_score: 0.8659.
RECM: 1600.3756623334853.


In [34]:
train_data = lgb.Dataset(X_train, label=y_train)
test_data = lgb.Dataset(X_test, label=y_test)

In [35]:
# modificando algunos de los parámetros del modelo que tuvo mejor rendimiento (model_gbm)

parameters = {
    'force_row_wise' : True,
    'is_unbalance' : True,
    'boosting'  : 'gbdt',
    'verbosity' : -1,
    'num_leaves' : 100,
    'learning_rate' : 0.5,
    'metric' : ['rmse', 'auc']}

In [36]:
callbacks = [lgb.early_stopping(10)]
t0 = time.time()
model_gbm = lgb.LGBMRegressor()
model_gbm.fit(X_train, y_train)
t1 = time.time()
total_time = t1 - t0
print(f'Tiempo total de entrenamiento:{total_time}')
predict = model_gbm.predict(X_test)
mse = mean_squared_error(predict, y_test)
rmse = np.sqrt(mse)
print(f'Métrica r2_score: {r2_score(y_test, predict).round(4)}.')
print(f'RECM: {rmse}.')

Tiempo total de entrenamiento:7.417747735977173
Métrica r2_score: 0.8445.
RECM: 1723.3256942827438.


<div class="alert alert-block alert-warning">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
Y aquí las implementaciones fueron similares buscando diferentes variables de los modelos  y viendo como estas variaciones impactan en las métricas, estupendo! Si quisieramos podríamos tratar de investigar el uso de alternativas de gridesearch para evaluar una gran cantidad de hiperparametros sin tener que repetir códigos.

Explicación:
GridSearchCV: Busca exhaustivamente en una cuadrícula de parámetros para encontrar la mejor combinación.
param_grid: Define los rangos de parámetros que quieres probar.
scoring: Puedes usar diferentes métricas de evaluación; en este caso, estamos utilizando el r2_score.
cv: Especifica el número de divisiones para la validación cruzada.
Este enfoque te permite evaluar una gran cantidad de combinaciones de hiperparámetros sin necesidad de escribir código repetitivo, y optimizarás los modelos para el conjunto de datos que estás usando.

In [ ]:
# Definir los hiperparámetros para SGDRegressor
# Definir el modelo
sgd = SGDRegressor()

# Definir el grid de hiperparámetros
param_grid_sgd = {
    'alpha': [0.0001, 0.001, 0.01, 0.1],
    'l1_ratio': [0.1, 0.5, 0.9],
    'max_iter': [1000, 2000, 3000],
    'learning_rate': ['constant', 'optimal', 'adaptive']
}

# Usar GridSearchCV para encontrar la mejor combinación de hiperparámetros
grid_search_sgd = GridSearchCV(estimator=sgd, param_grid=param_grid_sgd, cv=5, scoring='r2', n_jobs=-1)
grid_search_sgd.fit(X_train, y_train)

# Resultados
print("Best parameters found for SGDRegressor:")
print(grid_search_sgd.best_params_)
print("Best R2 score:", grid_search_sgd.best_score_)

/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_stochastic_gradient.py:1220: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn("Maximum number of iteration reached before "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_stochastic_gradient.py:1220: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn("Maximum number of iteration reached before "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_stochastic_gradient.py:1220: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn("Maximum number of iteration reached before "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_stochastic_gradient.py:1220: ConvergenceWarning: Maximum

/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_stochastic_gradient.py:1220: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn("Maximum number of iteration reached before "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_stochastic_gradient.py:1220: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn("Maximum number of iteration reached before "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_stochastic_gradient.py:1220: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn("Maximum number of iteration reached before "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_stochastic_gradient.py:1220: ConvergenceWarning: Maximum

In [ ]:
# # Definir el modelo
lgbm = lgb.LGBMRegressor()

# Definir el grid de hiperparámetros
param_grid_lgbm = {
    'num_leaves': [31, 50, 100],
    'learning_rate': [0.01, 0.05, 0.1],
    'n_estimators': [100, 200, 500],
    'boosting_type': ['gbdt', 'dart', 'goss']
}

# Usar GridSearchCV para encontrar la mejor combinación de hiperparámetros
grid_search_lgbm = GridSearchCV(estimator=lgbm, param_grid=param_grid_lgbm, cv=5, scoring='r2', n_jobs=-1)
grid_search_lgbm.fit(X_train, y_train)

# Resultados
print("Best parameters found for LGBMRegressor:")
print(grid_search_lgbm.best_params_)
print("Best R2 score:", grid_search_lgbm.best_score_)

In [ ]:
# Evaluar los modelos con los mejores parámetros
# Evaluar el mejor modelo SGD
best_sgd = grid_search_sgd.best_estimator_
y_pred_sgd = best_sgd.predict(X_test)
r2_sgd = r2_score(y_test, y_pred_sgd)
print(f"R2 score for best SGD model: {r2_sgd}")

# Evaluar el mejor modelo LGBM
best_lgbm = grid_search_lgbm.best_estimator_
y_pred_lgbm = best_lgbm.predict(X_test)
r2_lgbm = r2_score(y_test, y_pred_lgbm)
print(f"R2 score for best LGBM model: {r2_lgbm}")

## Conclusiones 

In [ ]:
# Datos de los modelos SGDL
sgdl_alpha = [0.01, 0.01, 0.01, 0.1]
sgdl_l1_ratio = [0.1, 1, 10, 0]
sgdl_r2_scores = [0.5608, 0.4166, 0.1028, 0.6432]
sgdl_times = [36.526, 36.513, 37.398, 9.402]

# Datos del modelo GBDT
gbdt_r2_scores = [0.8659, 0.8445]
gbdt_times = [8.184, 4.416]

# Configuración de la figura
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 6))

# Gráfico de r2_score
ax1.plot(sgdl_alpha, sgdl_r2_scores, marker='o', linestyle='-', color='b', label='SGDL')
ax1.axhline(y=max(gbdt_r2_scores), color='r', linestyle='--', label='GBDT Max r2_score')
ax1.set_title('Comparación de r2_score')
ax1.set_xlabel('Parámetro de Alpha (SGDL)')
ax1.set_ylabel('r2_score')
ax1.legend()

# Gráfico de tiempo de entrenamiento
ax2.plot(sgdl_alpha, sgdl_times, marker='o', linestyle='-', color='b', label='SGDL')
ax2.axhline(y=min(gbdt_times), color='r', linestyle='--', label='GBDT Min Tiempo')
ax2.set_title('Comparación de Tiempo de Entrenamiento')
ax2.set_xlabel('Parámetro de Alpha (SGDL)')
ax2.set_ylabel('Tiempo en Segundos')
ax2.legend()

plt.tight_layout()
plt.show()

En el análisis de los dos modelos, se pueden observar varias diferencias clave en términos de rendimiento y tiempo de entrenamiento:

1. Modelo de Regresión Lineal (SGDL):
Rendimiento (r2_score):
Con alpha=0.01, l1_ratio=0.1: 0.5608
Con alpha=0.01, l1_ratio=1: 0.4166
Con alpha=0.01, l1_ratio=10: 0.1028
Con alpha=0.1, l1_ratio=0: 0.6432 (el mejor de los modelos SGDL)
Tiempo de entrenamiento y predicción:
Varía de aproximadamente 9.4 a 37.4 segundos.
2. Modelo de Gradient Boosting (GBDT) usando LightGBM:
Rendimiento (r2_score):
En la primera prueba con num_boost_round=500: 0.8659 (mejor r2_score).
En la segunda prueba (parámetros ajustados): 0.8445.
Tiempo de entrenamiento y predicción:
Primera prueba: 8.18 segundos.
Segunda prueba: 4.42 segundos.
Error cuadrático medio (RMSE):
Primera prueba: 1600.38
Segunda prueba: 1723.33
- Comparación y Conclusiones:

Rendimiento: El modelo GBDT con LightGBM supera claramente al modelo SGDL en términos de r2_score, alcanzando un valor mucho más alto (0.8659 frente a 0.6432 en el mejor caso de SGDL). Esto indica que el modelo GBDT tiene una mejor capacidad para explicar la variabilidad de los datos.

Tiempo de entrenamiento: El modelo GBDT también se entrena y predice más rápido (alrededor de 4.42 a 8.18 segundos) en comparación con el mejor caso de SGDL, que tomó 9.4 segundos. Además, el rendimiento fue superior a pesar del menor tiempo de cómputo.

RMSE: Aunque el RMSE fue ligeramente peor en la segunda prueba de GBDT, el r2_score se mantuvo alto, lo que sugiere que el modelo aún generaliza bien.

Conclusión General: 
El modelo de Gradient Boosting (GBDT) utilizando LightGBM es claramente superior al modelo de regresión lineal SGDL en términos de rendimiento (r2_score) y tiempo de entrenamiento, por lo que sería la opción preferida.

<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
    
Jorge solo queda felicitarte porque te has destacado de gran forma en el proyecto, has cumplicado con todos los puntos importantes y has hecho agregados perfectos que solo revalidan tu gran capcaidad de código pero también tu gran capacidad de análisis frente a los resultados, felicitaciones!

# Lista de control

Escribe 'x' para verificar. Luego presiona Shift+Enter

- [x]  Jupyter Notebook está abierto
- [ ]  El código no tiene errores- [ ]  Las celdas con el código han sido colocadas en orden de ejecución- [ ]  Los datos han sido descargados y preparados- [ ]  Los modelos han sido entrenados
- [ ]  Se realizó el análisis de velocidad y calidad de los modelos